### ISSUE TITLE
Specifications for switch/swarm

### ISSUE THREAD

**Issue by Mikerah** [libp2p/specs#149]
I am currently going through the [js-libp2p-switch](https://github.com/libp2p/js-libp2p-switch) library in order to make a change needed in my gossipsub implementation. The library has a lot of moving parts and it isn't exactly straightforward on how each component works together.

Are there any specifications about the dialer machine that I can read about upon which this codebase is based on?

**Comment by Mikerah** [comment]
After going through all the files, I found [this short section](https://github.com/libp2p/specs/blob/master/8-implementations.md#81-swarm) in the implementation section. Is the dialer machine more of an implementation detail?

**Comment by raulk** [comment]
@jacobheun @vasco.santos

**Comment by jacobheun** [comment]
> Is the dialer machine more of an implementation detail?

Currently, yes. With the go dialer v2 coming up it will be even more different in how the implementations behave. There are some specifics that we should get documented however. For example, if the implementation supports PSK protection it should happen immediately after the connection is established, crypto should occur after that, followed by muxer selection, etc. We should spec out any implementation agnostic requirements that are needed for interop, and document specific implementation details in their respective repo. @raulk thoughts?

**Comment by yusefnapora** [comment]
I'm starting to make up a list of "things that need specs", and I've been thinking of this as part of an overall "Connection Establishment" spec, which would cover the whole handshake and connection upgrade process.

Are the implementation-agnostic pieces of the switch/swarm are all related to connection establishment, or are there other responsibilities that are visible "from the outside" and should be spec'd out?

**Comment by jacobheun** [comment]
> Are the implementation-agnostic pieces of the switch/swarm are all related to connection establishment, or are there other responsibilities that are visible "from the outside" and should be spec'd out?

I believe establishing the connection is the only agnostic component. Everything else would fall to the implementations to document.

**Comment by raulk** [comment]
Yeah, with dialer v2 we can provide a non-normative recommendation, but I don’t think it’s material for a normative spec.
